src/kudu/tserver/simple_tablet_manager.cc

```diff
@@ -117,36 +122,208 @@ TSTabletManager::TSTabletManager(TabletServer* server)
+      resolved_config.add_peers()->CopyFrom(peer);
+    } else {
+      LOG(INFO) << SecureShortDebugString(peer)
+                << " has no permanent_uuid. Determining permanent_uuid...";
+      RaftPeerPB new_peer = peer;
+      RETURN_NOT_OK_PREPEND(consensus::SetPermanentUuidForRemotePeer(server_->messenger(),
+                                                                     &new_peer),
+                            Substitute("Unable to resolve UUID for peer $0",
+                                       SecureShortDebugString(peer)));
+      resolved_config.add_peers()->CopyFrom(new_peer);
+    }
+  }
+
+  RETURN_NOT_OK(consensus::VerifyRaftConfig(resolved_config));
+  VLOG(1) << "Distributed Raft configuration: " << SecureShortDebugString(resolved_config);
+
+  *committed_config = resolved_config;
+  return Status::OK();
+}
+
+Status TSTabletManager::WaitUntilConsensusRunning(const MonoDelta& timeout) {
+  MonoTime start(MonoTime::Now());
+
+  int backoff_exp = 0;
+  const int kMaxBackoffExp = 8;
+  while (true) {
+    if (consensus_ && consensus_->IsRunning()) {
+      break;
+    }
+    MonoTime now(MonoTime::Now());
+    MonoDelta elapsed(now - start);
+    if (elapsed > timeout) {
+      return Status::TimedOut(Substitute("Raft Consensus is not running after waiting for $0:",
+                                         elapsed.ToString()));
+    }
+    SleepFor(MonoDelta::FromMilliseconds(1L << backoff_exp));
+    backoff_exp = std::min(backoff_exp + 1, kMaxBackoffExp);
+  }
+  return Status::OK();
+}
+
+Status TSTabletManager::WaitUntilRunning() {
+  TRACE_EVENT0("master", "SysCatalogTable::WaitUntilRunning");
+  int seconds_waited = 0;
+  while (true) {
+    Status status = WaitUntilConsensusRunning(MonoDelta::FromSeconds(1));
+    seconds_waited++;
+    if (status.ok()) {
+      LOG_WITH_PREFIX(INFO) << "configured and running, proceeding with master startup.";
+      break;
+    }
+    if (status.IsTimedOut()) {
+      LOG_WITH_PREFIX(INFO) <<  "not online yet (have been trying for "
+                               << seconds_waited << " seconds)";
+      continue;
+    }
+    // if the status is not OK or TimedOut return it.
+    return status;
+  }
+  return Status::OK();
+}
+
+Status TSTabletManager::InitRaftAsync(bool is_first_run) {
+  if (is_first_run) {
+    RETURN_NOT_OK(CreateNew(server_->fs_manager()));
+  } else {
+    RETURN_NOT_OK(Load(server_->fs_manager()));
+  }
+  return Status::OK();
+}
+
+bool TSTabletManager::IsInitialized() const {
+  return state() == MANAGER_RUNNING;
+}
+
+Status TSTabletManager::Init(bool is_first_run) {
   CHECK_EQ(state(), MANAGER_INITIALIZING);
 
-  InitLocalRaftPeerPB();
+  RETURN_NOT_OK_PREPEND(InitRaftAsync(is_first_run),
+                        "Failed to initialize raft async");
 
-  LOG(INFO) << LogPrefix(kSysCatalogTabletId) << "Bootstrapping tablet";
-  TRACE("Bootstrapping tablet");
+  // WaitUntilRunning() must run outside of the lock as to prevent
```

Comment:
Which lock are you referring to here? I got a bit confused.

Reply:
This comment is from kudu. I had left it as a reminder of concurrency issues which might be
TBD, but I will remove it to avoid confusion.